### PR TITLE
Support running multiple border routers on the same IP.

### PR DESCRIPTION
**Pull request by ercanucan**
This change enables deploying multiple border router instances
on the same IP by modifying the Ansible hostfile generation.

---
This change is [Reviewable](https://reviewable.io/reviews/netsec-ethz/scion-web/56)

### CODE REVIEW

**Comment by ercanucan** [comment]
@klausman @kormat

**Comment by kormat** [comment]
Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[ad_manager/models.py, line 344 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/56#-KdeiA8IzJLCrfojFByO:-KdeiA8IzJLCrfojFByP:bwrhf3a) ([raw file](https://github.com/netsec-ethz/scion-web/blob/e8fc5b13d6599521415e82033b250d7c8d44b81f/ad_manager/models.py#L344)):*
> ```Python
>     class Meta:
>         verbose_name = 'Router'
>         unique_together = (("ad", "addr", "interface_port"),)
> ```

This is also dependant on the overlay involved. It will work if the overlay is UDP/x, but it doesn't work for a plain IPv4 or IPv6 overlay.

---

*[ad_manager/util/hostfile_generator.py, line 68 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/56#-KdejvTOpEaug-BCoITN:-KdejvTOpEaug-BCoITO:bxbbi4x) ([raw file](https://github.com/netsec-ethz/scion-web/blob/e8fc5b13d6599521415e82033b250d7c8d44b81f/ad_manager/util/hostfile_generator.py#L68)):*
> ```Python
>     :param str isd_id: the ISD the service belongs to.
>     :param str as_id: the AS the service belongs to.
>     """
> ```

This can be simplified quite a bit. Roughly:

```
addr2instances = defaultdict(list)
add2name = defaultdict(list)

for name, addr in val:
  ...
  addr2instances[addr].append(instance_id)
  addr2name[addr].append(name)
```

---

*[ad_manager/util/hostfile_generator.py, line 87 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/56#-KdekW-reTmHvaXObh2W:-KdekW-reTmHvaXObh2X:b1w8s5a) ([raw file](https://github.com/netsec-ethz/scion-web/blob/e8fc5b13d6599521415e82033b250d7c8d44b81f/ad_manager/util/hostfile_generator.py#L87)):*
> ```Python
>     for addr, instances in addr2instances.items():
>         config[section_name][addr] = (base_tags + "'" + json.dumps(instances) +
>                                       "' # " + str(addr2name[addr]))
> ```

This method of string construction is really hard to follow.
```
"isd=%s as=%s instances='%s' # %s" % (isd_id, as_id, json.dumps(instances), addr2name[addr])
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-web/56)*

**Comment by ercanucan** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[ad_manager/models.py, line 344 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/56#-KdeiA8IzJLCrfojFByO:-Kdjer8E_z0Rm5hjh85S:b8mfi6o) ([raw file](https://github.com/netsec-ethz/scion-web/blob/e8fc5b13d6599521415e82033b250d7c8d44b81f/ad_manager/models.py#L344)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This is also dependant on the overlay involved. It will work if the overlay is UDP/x, but it doesn't work for a plain IPv4 or IPv6 overlay.
</blockquote></details>

Modified the constraint to be on the 'Addr' and 'Port', the main (public) address and the port of the router in the existing topo format). `models.py` needs a larger re-factoring along with a few other locations (front-end, back-end) to accommodate the changes that come with the new topology file format. Created an issue for that: https://github.com/netsec-ethz/scion-web/issues/57
(I found that the 'Port' field was missing from the original DB model, it was populated directly from the saved topology. Fixed that and also modified the other elements, for the sake of completeness in the current state of the code. )

---

*[ad_manager/util/hostfile_generator.py, line 68 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/56#-KdejvTOpEaug-BCoITN:-Kdjb5FRwrE4iP87xcrd:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/e8fc5b13d6599521415e82033b250d7c8d44b81f/ad_manager/util/hostfile_generator.py#L68)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This can be simplified quite a bit. Roughly:

```
addr2instances = defaultdict(list)
add2name = defaultdict(list)

for name, addr in val:
  ...
  addr2instances[addr].append(instance_id)
  addr2name[addr].append(name)
```
</blockquote></details>

Done.

---

*[ad_manager/util/hostfile_generator.py, line 87 at r1](https://reviewable.io:443/reviews/netsec-ethz/scion-web/56#-KdekW-reTmHvaXObh2W:-KdjblYkumYZOF-Mb7wQ:b-896fix) ([raw file](https://github.com/netsec-ethz/scion-web/blob/e8fc5b13d6599521415e82033b250d7c8d44b81f/ad_manager/util/hostfile_generator.py#L87)):*
<details><summary><i>Previously, kormat (Stephen Shirley) wrote…</i></summary><blockquote>

This method of string construction is really hard to follow.
```
"isd=%s as=%s instances='%s' # %s" % (isd_id, as_id, json.dumps(instances), addr2name[addr])
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-web/56)*

**Comment by kormat** [comment]
:lgtm:

---

Reviewed 3 of 4 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/netsec-ethz/scion-web/56#-:-Kdz98nLEJd5wn3DLOkU:bnfp4nl)*

**Comment by ercanucan** [comment]
Thanks!
